cli/maintainer/src/maintenance.rs

```diff
@@ -754,33 +772,75 @@ impl SolidoState {
 
-        let rewards = (reserve_st_sol - st_sol_amount).ok()?;
+        (reserve_st_sol - st_sol_amount).ok()
+    }
+
+    /// Try to sell the extra stSOL rewards for UST tokens or
+    /// to update the historical pool price exchange rate to protect us
+    /// against sandwiching attacks.
+    pub fn try_sell_anker_rewards(&self) -> Option<MaintenanceInstruction> {
+        let anker_state = self.anker_state.as_ref()?;
 
+        let rewards = self.get_anker_rewards()?;
         let min_rewards_to_sell = self
             .solido
             .exchange_rate
             .exchange_sol(Self::MINIMUM_WITHDRAW_AMOUNT)
             .expect("The price of a signature should be small enough that it doesn't overflow.");
 
-        // Estimate the amount of UST that we will get when swapping against the constant product
-        // pool. This is only an approximation because it doesn’t uphold the constant product
-        // invariant, but this is a reasonable approximation when the rewards to sell are small
-        // compared to the pool balance. Also, this ignores swap fees. It would have been nice
-        // to use the actual `spl_token_swap::state::SwapV1` instance, but it is not `Send`, so
-        // we can’t store it in our state.
-        let ust_per_st_sol = Rational {
-            numerator: anker_state.pool_ust_balance.0,
-            denominator: anker_state.pool_st_sol_balance.0,
-        };
-        let expected_proceeds = (rewards * ust_per_st_sol)
-            .map(|x| MicroUst(x.0))
-            .unwrap_or(MicroUst(0));
+        let expected_st_sol_price_in_ust = get_one_st_sol_for_ust_price_from_pool(
+            &anker_state.constant_product_calculator,
+            &anker_state.pool_st_sol_account,
+            &anker_state.pool_ust_account,
+            anker_state.pool_st_sol_balance,
+            anker_state.pool_ust_balance,
+        )
+        .ok()?;
+
+        let expected_proceeds = MicroUst(rewards.0 * expected_st_sol_price_in_ust.0);
         // We want at least 0.01 UST out if we are going to do the swap at all.
         let min_proceeds = MicroUst(10_000);
 
+        // Check if we can sell the rewards with the preset slippage tolerance.
+        // Note that this might change when the instruction gets included in the block.
+        // TODO(#539): Try to arbitrage in case this condition does not hold.
+        let minimum_ust_amount_for_rewards = anker_state
+            .anker
+            .historical_st_sol_prices
+            .minimum_ust_swap_amount(rewards, anker_state.anker.sell_rewards_min_out_bps)
+            .ok()?;
+
         // We should not call the instruction if the rewards are 0, or if the rewards are so small
         // that the transaction cost is a significant portion of the rewards.
-        if rewards < min_rewards_to_sell || expected_proceeds < min_proceeds {
+        if rewards < min_rewards_to_sell {
```

Comment:
We can move this check all the way to the top right after the definition of `min_rewards to sell`: if we don’t have enough, we don’t need to bother with any of the rest.

(I don’t have a strong opinion on this, but now that I’m reading this function front to back, I think it would be a bit easier to follow if it’s a list of check->action, instead of computing various values first and then having all the checks at the end.)

Reply:
Yes! That's much better, thanks!